### PR TITLE
feat: build operator image with chainguard

### PR DESCRIPTION
**Pull request by emosbaugh**
Builds images with chainguard (melange and apko)

https://github.com/replicatedhq/embedded-cluster-operator/releases/tag/v0.40.2-build.1

### CODE REVIEW

File: .github/workflows/deploy-helm-production.yaml [hunk 1]
```diff
@@ -20,21 +21,50 @@ jobs:
...
+        path: bin/
+    - name: Build melange package
+      run: |
+        export VERSION=${{steps.tag.outputs.tag}}
+        make melange
+    - name: Publish apko image
+      run: |
+        export VERSION=${{steps.tag.outputs.tag}}
+        export IMAGE=registry.replicated.com/library/embedded-cluster-operator-image:${VERSION}
+        make apko-login \
```
Comment (sgalsaleh) [inline .github/workflows/deploy-helm-production.yaml:58]: why not use their melange/apko github actions like we do in kots?

Comment (emosbaugh) [reply .github/workflows/deploy-helm-production.yaml:58]: This way i can unify how we build and test in the dev environment. It makes for a better developer experience in my opinion.

File: deploy/melange.tmpl.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      - https://packages.wolfi.dev/os
+    keyring:
+      - https://packages.wolfi.dev/os/wolfi-signing.rsa.pub
+    packages:
+      - busybox
+
+pipeline:
+  - runs: |
+      set -ex
+      cp manager "${{targets.destdir}}/manager"
```
Comment (sgalsaleh) [inline deploy/melange.tmpl.yaml:21]: why not build the binary here? it's more portable that way and seems like the established pattern

Comment (emosbaugh) [reply deploy/melange.tmpl.yaml:21]: i agree. ill fix this

File: .github/workflows/deploy-helm-production.yaml (hunk 1 above)
Comment (sgalsaleh) [reply .github/workflows/deploy-helm-production.yaml:58]: yeah good point

File: Makefile [hunk 3]
```diff
@@ -311,3 +311,53 @@ build-chart-ttl.sh: export OPERATOR_IMAGE_TAG = 24h
 build-chart-ttl.sh: export CHART_REMOTE = oci://ttl.sh/${CURRENT_USER}
 build-chart-ttl.sh:
 	cd charts/embedded-cluster-operator && ../../scripts/publish-helm-chart.sh
+
+.PHONY: apko-build
+apko-build: export IMAGE = ttl.sh/${CURRENT_USER}/embedded-cluster-operator-image:24h
```
Comment (sgalsaleh) [inline Makefile:316]: wouldn't this always override the environment variables?

File: .github/workflows/deploy-helm-production.yaml (hunk 1 above)
Comment (St0rmz1) [reply .github/workflows/deploy-helm-production.yaml:58]: However, make sure we're still generating the SBOM. Their melange/apko github actions will auto generate and publish SBOMs with any images. And then we need that to generate the SLSA attestations.

File: Makefile (hunk 3 above)
Comment (emosbaugh) [reply Makefile:316]: yes. fixing

File: .github/workflows/deploy-helm-production.yaml (hunk 1 above)
Comment (emosbaugh) [reply .github/workflows/deploy-helm-production.yaml:58]: Their action runs the same command we do. Running the command i see the sbom being pushed to the docker registry with a tag

https://github.com/chainguard-images/actions/blob/main/apko-publish/action.yaml#L173

Comment (emosbaugh) [reply .github/workflows/deploy-helm-production.yaml:52]:
```2024/07/16 21:34:05 INFO built index file as /tmp/apko-temp-2007370370/index.json
2024/07/16 21:34:21 INFO publishing index tag ttl.sh/ethan/embedded-cluster-operator-image:24h
2024/07/16 21:34:26 INFO Published SBOM ttl.sh/ethan/embedded-cluster-operator-image:sha256-5e44d7c393718b5c1d94997a6d28e260440ac2a976cb66fe73321368d5cd317c.sbom
2024/07/16 21:34:30 INFO Published SBOM ttl.sh/ethan/embedded-cluster-operator-image:sha256-6cca2104395e0861fb0d2047cae91581d2e4039d3e2d8a517428f437341d9536.sbom
ttl.sh/ethan/embedded-cluster-operator-image@sha256:6cca2104395e0861fb0d2047cae91581d2e4039d3e2d8a517428f437341d9536```

File: .github/workflows/deploy-helm-production.yaml [hunk 4]
```diff
@@ -63,18 +87,13 @@ jobs:
...
-      - name: Set up Go
-        uses: actions/setup-go@v5
+      - name: Download artifact
+        uses: actions/download-artifact@v4
         with:
-          go-version-file: go.mod
-      - name: Build linux-amd64
-        run: |
-          make build
+          name: manager
```
Comment (sgalsaleh) [inline .github/workflows/deploy-helm-production.yaml:95]: hmm, thinking about this, the binary in the image will be CVE free, but this is the one that will end up in production and used by kots. i worked around a similar situation with the schemahero binary and the kotsadm-migrations image in kots by extracting the binary from the image published by apko: https://github.com/replicatedhq/kots/blob/main/.github/actions/build-custom-melange-package/action.yml#L44-L45

File: .github/workflows/deploy-helm-production.yaml (hunk 1 above)
Comment (St0rmz1) [reply .github/workflows/deploy-helm-production.yaml:35]: Awesome. thank you for checking!

File: .github/workflows/deploy-helm-production.yaml (hunk 4 above)
Comment (emosbaugh) [reply .github/workflows/deploy-helm-production.yaml]: changed

File: Makefile [hunk 5]
```diff
@@ -311,3 +314,58 @@ build-chart-ttl.sh: export OPERATOR_IMAGE_TAG = 24h
...
+	docker run -v "${PWD}":/work -v "${PWD}"/build/.docker:/root/.docker -w /work/build \
+		cgr.dev/chainguard/apko login -u "${USERNAME}" \
+			--password "${PASSWORD}" "${REGISTRY}"
+
+.PHONY: melange
+melange: export ARCHS ?= amd64
+melange: melange-template
+	mkdir -p build
+	for f in pkg controllers main.go go.mod go.sum Makefile ; do \
+		rm -rf "build/$$f" && cp -r $$f build/ ; \
```
Comment (sgalsaleh) [inline Makefile:345]: should we add `build/` to .gitignore?

Comment (emosbaugh) [reply Makefile:345]: https://github.com/replicatedhq/embedded-cluster-operator/pull/250/files#diff-bc37d034bad564583790a46f19d807abfe519c5671395fd494d8cce506c42947R24

**Review by sgalsaleh — APPROVED** [review]

File: Makefile (hunk 5 above)
Comment (sgalsaleh) [reply Makefile:345]: oh, totally missed it 🤦

**Pull request merged**